File: motia-js/playground/steps/api-steps/sse.step.ts
import { type Handlers, http, type StepConfig } from 'motia'

export const config = {
  name: "SSE Example",
  description: "Accepts form-data and streams back random items as SSE",
  flows: ["sse-example"],
  triggers: [http("POST", "/sse")],
  enqueues: [],
} as const satisfies StepConfig

export const handler: Handlers<typeof config> = async ({ request, response }, { logger }) => {
  logger.info('FormData received', { headers: request.headers })
  response.status(200)
  response.headers({ 
    'content-type': 'text/event-stream',
    'cache-control': 'no-cache',
    connection: 'keep-alive',
  })
  logger.info('Headers set', { headers: response.headers })

  const responses: string[] = []

  for await (const chunk of request.requestBody.stream) {
    responses.push(Buffer.from(chunk).toString('binary'))
  }

  const parts: Record<string, string> = {}
  
  responses
    .filter(s => typeof s === 'string' && s.trim().length > 0)
    .forEach((s) => {
      const pairs = s.split('&')
      for (const pair of pairs) {
        const [key, value] = pair.split('=')
        if (key) {
          parts[decodeURIComponent(key)] = value ? decodeURIComponent(value) : ''
        }
      }
    })

  const items = generateRandomItems(parts)

  for (const item of items) {
    response.stream.write(`event: item\ndata: ${JSON.stringify(item)}\n\n`)
    await sleep(300 + Math.random() * 700)
  }

  response.stream.write(`event: done\ndata: ${JSON.stringify({ total: items.length })}\n\n`)
  response.close()
}

function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

type FormField = { name: string; value: string }

function generateRandomItems(parts: Record<string, string>) {
  const fields: FormField[] = Object.entries(parts).map(([name, value]) => ({ name, value }))
  const count = 5 + Math.floor(Math.random() * 6)
  const adjectives = ['swift', 'lazy', 'bold', 'calm', 'fierce', 'gentle', 'sharp', 'wild']
  const nouns = ['falcon', 'river', 'mountain', 'crystal', 'thunder', 'shadow', 'ember', 'frost']

  return Array.from({ length: count }, (_, i) => ({
    id: `item-${Date.now()}-${i}`,
    label: `${adjectives[Math.floor(Math.random() * adjectives.length)]} ${nouns[Math.floor(Math.random() * nouns.length)]}`,
    score: Math.round(Math.random() * 100),
    source: fields.length > 0 ? fields[i % fields.length] : null,
  }))
}

File: motia-py/playground/steps/api_steps/sse_step.py
"""Accepts multipart/form-data and streams back random items as SSE."""

import asyncio
import json
import math
import random
import time
from typing import Any
from urllib.parse import unquote

from motia import ApiStreamRequest, FlowContext, http

config = {
    "name": "SSE Example",
    "description": "Accepts form-data and streams back random items as SSE",
    "flows": ["sse-example"],
    "triggers": [
        http("POST", "/sse"),
    ],
    "enqueues": [],
}


async def handler(args: ApiStreamRequest[Any], ctx: FlowContext[Any]) -> None:
    """Read form body, then stream random items back as SSE."""
    request = args.request
    response = args.response

    ctx.logger.info("FormData received", {"headers": request.headers})

    await response.status(200)
    await response.headers({
        "content-type": "text/event-stream",
        "cache-control": "no-cache",
        "connection": "keep-alive",
    })

    raw_chunks: list[str] = []
    async for chunk in request.request_body.stream:
        if isinstance(chunk, bytes):
            raw_chunks.append(chunk.decode("latin-1"))
        else:
            raw_chunks.append(str(chunk))

    parts: dict[str, str] = {}
    for s in raw_chunks:
        s = s.strip()
        if not s:
            continue
        for pair in s.split("&"):
            if "=" in pair:
                key, value = pair.split("=", 1)
                parts[unquote(key)] = unquote(value)
            elif pair:
                parts[unquote(pair)] = ""

    items = _generate_random_items(parts)

    for item in items:
        response.writer.stream.write(f"event: item\ndata: {json.dumps(item)}\n\n".encode("utf-8"))
        await asyncio.sleep(0.3 + random.random() * 0.7)

    response.writer.stream.write(f"event: done\ndata: {json.dumps({'total': len(items)})}\n\n".encode("utf-8"))
    response.close()


def _generate_random_items(parts: dict[str, str]) -> list[dict[str, Any]]:
    fields = [{"name": k, "value": v} for k, v in parts.items()]
    count = 5 + math.floor(random.random() * 6)

    adjectives = ["swift", "lazy", "bold", "calm", "fierce", "gentle", "sharp", "wild"]
    nouns = ["falcon", "river", "mountain", "crystal", "thunder", "shadow", "ember", "frost"]

    return [
        {
            "id": f"item-{int(time.time() * 1000)}-{i}",
            "label": f"{random.choice(adjectives)} {random.choice(nouns)}",
            "score": round(random.random() * 100),
            "source": fields[i % len(fields)] if fields else None,
        }
        for i in range(count)
    ]
